source/common/http/http2/codec_impl.cc

```diff
@@ -90,7 +90,19 @@ void ConnectionImpl::StreamImpl::encode100ContinueHeaders(const HeaderMap& heade
 
 void ConnectionImpl::StreamImpl::encodeHeaders(const HeaderMap& headers, bool end_stream) {
   std::vector<nghttp2_nv> final_headers;
-  buildHeaders(final_headers, headers);
+
+  if (Http::Utility::isUpgrade(headers)) {
+    HeaderMapImpl modified_headers(headers);
+    upgrade_type_ = headers.Upgrade()->value().c_str();
+    if (headers.Status()) {
+      Http::Utility::transformUpgradeResponseFromH1toH2(modified_headers);
+    } else {
+      Http::Utility::transformUpgradeRequestFromH1toH2(modified_headers);
+    }
+    buildHeaders(final_headers, modified_headers);
+  } else {
+    buildHeaders(final_headers, headers);
+  }
```

Comment:
Is there a way to add some indication on the direction? whether the response is being returned to downstream or request is being forwarded to upstream ? that should make this code a bit more clear.